### PR TITLE
Fix for JSIP-363: read bytes in UTF8 from input buffer

### PR DESCRIPTION
**Pull request by fre42**
UTF8 characters are not handled correctly when receiving a SIP message via TCP or TLS under MS Windows 7.
The problem is caused by the code in PipelinedMsgParser.java in the run() method near line 462.
The bytes are read from buffer inputBuffer.toString().getBytes() without specifying the charset which needs to be UTF8.
This does only work correctly when your local machine is configured to use UTF-8 charset by default.
For MS Windows machines this is not the default.

I would suggest to change the call to inputBuffer.toString().getBytes("UTF-8") and do some error handling for the (strange) case that there is no UTF-8 charset available.

The issue was first reported in https://java.net/jira/browse/JSIP-363. In usnistgov/jsip it has been fixed in a similar way.

### CODE REVIEW

File: src/gov/nist/javax/sip/parser/PipelinedMsgParser.java [hunk 1]
```diff
@@ -459,7 +460,14 @@ public void run() {
                     if (stackLogger.isLoggingEnabled(StackLogger.TRACE_DEBUG)) {
                         stackLogger.logDebug("About to parse : " + inputBuffer.toString());
                     }
-                    sipMessage = smp.parseSIPMessage(inputBuffer.toString().getBytes(), false, false, sipMessageListener);
+                    byte[] inputBufferBytes;
+                    try {
+                        inputBufferBytes = inputBuffer.toString().getBytes("UTF-8");
+                    } catch (UnsupportedEncodingException e) {
+                        // log the lack of UTF8 support here?
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/parser/PipelinedMsgParser.java]: i think we should log this under WARNING level, this means an incorrect JVM configuration...

**Comment by fre42** [comment]
Yes, OK. When UTF-8 is really missing on a misconfigured system then the fallback code after _catch_ would work. But in this case you would see the WARN message for every received packet. That's the reason why I was not sure to place logging here.

**Comment by jaimecasero** [comment]
@fre42 printing a WARN message for every packet is heavy, but a system running with incorrect encoding configuration is quite unpredictable anyway. If the option is to silently go ahead, then I prefer printing everytime.

 Another solution would be to test for encoding on stack creation and print the warning there. Would you like to contribute that along with this PR?

**Comment by fre42** [comment]
I've added some code on stack creation which tests if "UTF-8" is available and which logs a WARN if not.

File: src/gov/nist/javax/sip/parser/PipelinedMsgParser.java [hunk 2]
```diff
@@ -464,7 +464,7 @@ public void run() {
                     try {
                         inputBufferBytes = inputBuffer.toString().getBytes("UTF-8");
                     } catch (UnsupportedEncodingException e) {
-                        // log the lack of UTF8 support here?
+                        // fall back to default encoding. The lack of UTF8 support has been logged at SIP stack startup
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/parser/PipelinedMsgParser.java]: great comment!!! thanks

File: src/gov/nist/javax/sip/SipStackImpl.java [hunk 3]
```diff
@@ -694,6 +696,14 @@ protected SipStackImpl() {
 		this.eventScanner = new EventScanner(this);
 		this.listeningPoints = new Hashtable<String, ListeningPointImpl>();
 		this.sipProviders = Collections.synchronizedList(new LinkedList<SipProviderImpl>());
+		try {
+			Charset charset = Charset.forName("UTF-8");
+			if (charset == null) {
+				throw new UnsupportedCharsetException("Unsupported charset UTF-8");
+			}
+		} catch (Exception e) {
+			logger.logWarning("UTF-8 charset cannot be used this system. This will lead to unpredictable behavior when parsing SIP messages: " + e.getMessage());
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/SipStackImpl.java]: this is what we were looking for

**Pull request merged**